### PR TITLE
Allow ALTER TABLE ADD COLUMN with DEFAULT on compressed hypertable

### PR DESCRIPTION
**Pull request by svenklemm**
When adding a new column with a default postgres doesnt materialize
the default expression but just stores it in catalog with a flag
signaling that it is missing from the physical row.
Any expressions used as default that do not require materialization
can be allowed on compressed hypertables as well.

The following statements will work on compressed hypertables with
this patch:
ALTER TABLE t ADD COLUMN c1 int DEFAULT 42;
ALTER TABLE t ADD COLUMN c2 text NOT NULL DEFAULT 'abc';

### CODE REVIEW

File: tsl/test/expected/compression_ddl.out [hunk 1]
```diff
@@ -980,3 +980,39 @@ WHERE hypertable_name = 'test1' and attname = 'bigintcol' ;
...
+
+ALTER TABLE test_defaults ADD COLUMN c1 int;
+ALTER TABLE test_defaults ADD COLUMN c2 int NOT NULL DEFAULT 42;
+SELECT * FROM test_defaults ORDER BY 1;
+             time             | device_id | c1 | c2 
+------------------------------+-----------+----+----
+ Sat Jan 01 00:00:00 2000 PST |         1 |    | 42
+ Mon Jan 01 00:00:00 2001 PST |         1 |    | 42
+(2 rows)
+
```
Comment (gayyappan) [inline tsl/test/expected/compression_ddl.out:1018]: Need test with insert into compressed chunk a) with a value for the new column, b) without  a value for the new column.
Followed by a recompress_chunk to ensure that works as expected.

File: tsl/test/sql/compression_ddl.sql [hunk 2]
```diff
@@ -465,5 +465,20 @@ ALTER TABLE i2844 SET (timescaledb.compress = FALSE);
 -- compression alter schema tests
 \ir include/compression_alter.sql
 
-
+-- test compression default handling
```
Comment (gayyappan) [inline tsl/test/sql/compression_ddl.sql:468]: The alter tests are in include/compression_alter.sql. Could you please move this test to that file?

File: tsl/test/expected/compression_ddl.out (hunk 1 above)
Comment (svenklemm) [reply tsl/test/expected/compression_ddl.out:1018]: done

File: tsl/test/sql/compression_ddl.sql (hunk 2 above)
Comment (svenklemm) [reply tsl/test/sql/compression_ddl.sql]: changed

**Review by nikkhils — APPROVED** [review]
LGTM

**Review by mfundul — APPROVED** [review]
The below error messages probably need to be rephrased:
```
errmsg("cannot add column with constraints "
"to a hypertable that has compression enabled")
```
```
errmsg("cannot add column with constraints or defaults to a hypertable that has "
"compression enabled")
```

File: tsl/test/expected/compression_errors.out [hunk 3]
```diff
@@ -321,23 +321,21 @@ alter table table_constr add constraint table_constr_fk_add_after FOREIGN KEY(d)
...
 -- ddl ADD column variants that are not supported
 ALTER TABLE table_constr ADD COLUMN newcol integer CHECK ( newcol < 10 );
-ERROR:  cannot add column with constraints or defaults to a hypertable that has compression enabled
+ERROR:  cannot add column with constraints to a hypertable that has compression enabled
 ALTER TABLE table_constr ADD COLUMN newcol integer UNIQUE;
-ERROR:  cannot add column with constraints or defaults to a hypertable that has compression enabled
+ERROR:  cannot add column with constraints to a hypertable that has compression enabled
 ALTER TABLE table_constr ADD COLUMN newcol integer PRIMARY KEY;
-ERROR:  cannot add column with constraints or defaults to a hypertable that has compression enabled
+ERROR:  cannot add column with constraints to a hypertable that has compression enabled
```
Comment (mkindahl) [inline tsl/test/expected/compression_errors.out:328]: I do not see the old error triggered anywhere, but it is still in the code. Do you have a test case for that? I don't see one.

**Review by mkindahl — COMMENTED** [review]
Just two comments, both about test coverage.

File: tsl/src/nodes/decompress_chunk/exec.c [hunk 4]
```diff
@@ -410,7 +410,13 @@ decompress_chunk_create_tuple(DecompressChunkState *state)
...
 					AttrNumber attr = AttrNumberGetAttrOffset(column->attno);
 
-					if (column->compressed.iterator != NULL)
+					if (!column->compressed.iterator)
+					{
+						slot->tts_values[attr] = getmissingattr(slot->tts_tupleDescriptor,
+																attr + 1,
+																&slot->tts_isnull[attr]);
+					}
+					else
```
[the remark this answers is not in the document]
Comment (svenklemm) [reply tsl/src/nodes/decompress_chunk/exec.c:419]: This does not do what you claim it does. Iterator works on batches and if you do an insert you create a new batch of size 1.

**Pull request merged**